File: clients/dotnet/tests/Novamem.Tests/RouteTests.cs
#nullable enable

using System.IO;
using System.Text.Json;
using Xunit;

namespace Novamem.Tests;

public class RouteTests
{
    // proved by: renaming Client.SessionRecapAsync fails this test (and the
    // generated dispatch table's compilation).
    [Fact]
    public void TestEveryRouteIsAccounted()
    {
        var routes = JsonDocument
            .Parse(File.ReadAllText(Path.Combine(ScenarioServer.Contract(), "routes.json")))
            .RootElement;
        foreach (var route in routes.EnumerateObject())
        {
            if (!route.Value.TryGetProperty("methods", out var ms))
                continue;
            foreach (var m in ms.EnumerateArray())
            {
                var parts = m.GetProperty("name").GetString()!.Split('.');
                var type = typeof(Client).Assembly.GetType("Novamem." + parts[0])!;
                Assert.True(
                    type.GetMethod(parts[1] + "Async") is not null,
                    $"{route.Name} {parts[0]}.{parts[1]}Async"
                );
                Assert.True(
                    Dispatch.Table.ContainsKey(m.GetProperty("name").GetString()!),
                    $"{route.Name}: no dispatch entry"
                );
            }
        }
    }
}

File: clients/swift/Tests/NovamemTests/RouteTests.swift
import Foundation
import XCTest

final class RouteTests: XCTestCase {
    /// proved by: deleting a dispatch entry fails this test; renaming
    /// Client.sessionRecap fails compilation of the dispatch table.
    func testEveryRouteIsAccounted() throws {
        let url = URL(fileURLWithPath: #filePath).deletingLastPathComponent().appendingPathComponent("../../../contract/routes.json")
            .standardizedFileURL
        let routes = try XCTUnwrap(try JSONSerialization.jsonObject(with: Data(contentsOf: url)) as? [String: [String: Any]])
        for (key, r) in routes {
            for m in (r["methods"] as? [[String: Any]]) ?? [] {
                let name = try XCTUnwrap(m["name"] as? String)
                XCTAssertNotNil(dispatchTable[name], "\(key): no dispatch entry for \(name)")
            }
        }
    }
}
